### DISCUSSION TITLE
Completion list when label of items starts with hyphen(dash)

### DISCUSSION THREAD

**Discussion by yangwooseong** [microsoft/vscode-discussions#1006] [Extension Development QnA] [answered]
I am creating an extension that autocompletes. Why does filtering not work when the item label starts with -? It doesn't work only in TypeScript files, but it works well in CSS files. Maybe should I report this in issue? it looks there is somewhat similar issue (https://github.com/microsoft/vscode/issues/188162) but not exactly the same.

<img width="312" alt="스크린샷 2024-01-10 오후 10 03 48" src="https://github.com/microsoft/vscode-discussions/assets/28595102/e4851a84-7f54-4e0b-8f31-bff5991a5b47">

<img width="689" alt="image" src="https://github.com/microsoft/vscode-discussions/assets/28595102/213d157f-7a9b-4637-9e78-20393911169e">

my code is following:

```typescript
connection.onComplete(() => {
  return [{
    detail: '#00000008',
    filterText: '-b-black-3',
    insertText: '-b-black-3',
    kind: CompletionItemKind.Color,
    label:  '-b-black-3',
 }, ... ]
})
```

Also, does color preview not work when it is rgba value? It does not show color preview like below.

<img width="591" alt="image" src="https://github.com/microsoft/vscode-discussions/assets/28595102/5a90659e-d977-47a2-bb44-5009c7f30a2a">

**Comment by isidorn** [comment]
@jrieken might be able to help here.

**Comment by jrieken** [comment, marked as answer]
A completion item provider is invoked, returns completions but they don't show up! 😕  What's the issue? Filtering! Filtering compares completions with a prefix and hides those that don't match.

#### Example

* trigger completions here: `say.he|` (`|` is the cursor)
* provider returns `abc`, `help`, `hood`
* the editor takes the `he`-prefix and scores it against `abc`, `help`, and `hood`. 
* scoring yields these result: 
  * `abc` - no match, don't show
  *  `help` - good match, show
  * `hood`  - no match, only `h` occurs, don't show
* editor shows `help` suggestion, does *not* show `abc` and `hood` suggestion

#### Extensions Rule!

The above sample was using "defaults" to determine the prefix and the filter criteria. 

 * `prefix:`  each completion can define its very own prefix via [`vscode.CompletionItem#range`](https://github.com/microsoft/vscode/blob/47b1183dfda519c32d78b591cfa721d3b0b874ae/src/vs/vscode.d.ts#L3803). Taking the example above, valid prefixes can be `say.he` or `.he` or ` ` (the empty string). Any range is valid as long as it includes the cursor position and is single line only.
* `filter:` each completion can define what string to use when filtering. The default is the label but via [`vscode.CompletionItem#filterText`](https://github.com/microsoft/vscode/blob/47b1183dfda519c32d78b591cfa721d3b0b874ae/src/vs/vscode.d.ts#L3776) any string can be defined

When to use what? It depends on the language, its word definition, and if you control/understand the word definition or not. The defaults usually work well enough but some languages might use a non-intuitive word definition, or completions span multiple words, or completions include non-word character etc. In all those cases the `range`-property should be used.

**Reply by yangwooseong** [reply]
I didn't understand what filtering had to do with how it works differently in css and ts. Thanks anyway
